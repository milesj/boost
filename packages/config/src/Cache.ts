import fs from 'fs';
import { Path } from '@boost/common';

export interface FileCache<T> {
	content: T;
	exists: boolean;
	mtime: number;
}

export default class Cache {
	configDir?: Path;

	dirFilesCache: Record<string, Path[]> = {};

	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	fileContentCache: Record<string, FileCache<any>> = {};

	pkgPath?: Path;

	rootDir?: Path;

	async cacheFileContents<T>(path: Path, cb: () => Promise<T>): Promise<T> {
		const key = path.path();
		const cache = this.fileContentCache[key];
		const stats = await fs.promises.stat(path.path());

		if (cache && cache.mtime === stats.mtimeMs) {
			return cache.content;
		}

		const content = await cb();

		this.fileContentCache[key] = {
			content,
			exists: true,
			mtime: stats.mtimeMs,
		};

		return content;
	}

	async cacheFilesInDir(dir: Path, cb: () => Promise<Path[]>): Promise<Path[]> {
		const key = dir.path();

		if (this.dirFilesCache[key]) {
			return this.dirFilesCache[key];
		}

		const files = await cb();

		this.dirFilesCache[key] = files;

		return files;
	}

	clearFileCache() {
		this.fileContentCache = {};
	}

	clearFinderCache() {
		this.dirFilesCache = {};
	}

	getFileCache<T>(path: Path): FileCache<T> | null {
		return this.fileContentCache[path.path()] || null;
	}

	markMissingFile(path: Path): this {
		this.fileContentCache[path.path()] = {
			content: null,
			exists: false,
			mtime: 0,
		};

		return this;
	}
}
