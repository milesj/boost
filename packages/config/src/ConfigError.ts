import { createScopedError } from '@boost/internal';

const errors = {
	EXTENDS_ONLY_ROOT: 'Extends setting `{0}` must only be defined in a root config.',
	EXTENDS_UNKNOWN_PATH: 'Cannot extend configuration. Unknown module or file path "{0}".',
	LOADER_UNSUPPORTED: 'Unsupported loader format "{0}".',
	PACKAGE_UNKNOWN_SCOPE: 'Unable to determine package scope. No parent `package.json` found.',
	ROOT_INVALID: 'Invalid configuration root. Requires a `{0}` folder and `package.json`.',
	ROOT_NO_PACKAGE:
		'Config folder `{0}` found without a relative `package.json`. Both must be located in the project root.',
	ROOT_ONLY_OVERRIDES: 'Overrides setting `{0}` must only be defined in a root config.',
};

export type ConfigErrorCode = keyof typeof errors;

export const ConfigError = createScopedError<ConfigErrorCode>('CFG', 'ConfigError', errors);
