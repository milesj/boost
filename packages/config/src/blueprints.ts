import { predicates } from '@boost/common';
import { ExtendsSetting, PluginsSettingOption } from './types';

const { array, bool, object, string, union } = predicates;

export const extendsSetting = union<ExtendsSetting>(
  [string().notEmpty(), array(string().notEmpty())],
  [],
);

export const pluginsSetting = object(union<PluginsSettingOption>([bool(), object()], {}));
