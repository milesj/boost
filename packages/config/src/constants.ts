import { ExtType } from './types';

export const CONFIG_FOLDER = '.config';

export const PACKAGE_FILE = 'package.json';

export const DEFAULT_EXTS: ExtType[] = ['js', 'json', 'cjs', 'mjs', 'ts', 'json5', 'yaml', 'yml'];
