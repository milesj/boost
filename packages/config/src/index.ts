/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export * from './ConfigError';
export * from './Configuration';
export * from './helpers/getEnv';
export * from './helpers/mergeArray';
export * from './helpers/mergeExtends';
export * from './helpers/mergeObject';
export * from './helpers/mergePlugins';
export * from './helpers/overwrite';
export * from './schemas';
export * from './types';
