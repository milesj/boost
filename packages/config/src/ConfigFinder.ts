/* eslint-disable no-param-reassign, no-await-in-loop */

import {
  Predicates,
  Path,
  PackageStructure,
  toArray,
  isModuleName,
  isFilePath,
} from '@boost/common';
import { RuntimeError } from '@boost/internal';
import minimatch from 'minimatch';
import loadCjs from './loaders/cjs';
import loadJs from './loaders/js';
import loadJson from './loaders/json';
import loadMjs from './loaders/mjs';
import loadYaml from './loaders/yaml';
import Finder from './Finder';
import createFileName from './helpers/createFileName';
import getEnv from './helpers/getEnv';
import { CONFIG_FOLDER, DEFAULT_EXTS, PACKAGE_FILE } from './constants';
import {
  ConfigFile,
  ConfigFinderOptions,
  ExtendsSetting,
  ExtType,
  OverridesSetting,
} from './types';

export default class ConfigFinder<T extends object> extends Finder<
  ConfigFile<T>,
  ConfigFinderOptions<T>
> {
  blueprint({ array, bool, func, shape, string }: Predicates) {
    return {
      extendsSetting: string(),
      extensions: array(string<ExtType>(), DEFAULT_EXTS),
      includeEnv: bool(true),
      loaders: shape({
        cjs: func(loadCjs).notNullable(),
        js: func(loadJs).notNullable(),
        json: func(loadJson).notNullable(),
        mjs: func(loadMjs).notNullable(),
        yaml: func(loadYaml).notNullable(),
        yml: func(loadYaml).notNullable(),
      }).exact(),
      name: string()
        .required()
        .camelCase(),
      overridesSetting: string(),
    };
  }

  /**
   * Determine a files package scope by finding the first parent `package.json`
   * when traversing up directories. We will leverage the cache as much as
   * possible for performance.
   *
   * @see https://nodejs.org/api/esm.html#esm_package_scope_and_file_extensions
   */
  async determinePackageScope(dir: Path): Promise<PackageStructure> {
    let currentDir = dir.isDirectory() ? dir : dir.parent();

    while (!this.isFileSystemRoot(currentDir)) {
      const pkgPath = currentDir.append(PACKAGE_FILE);
      const cache = this.cache.getFileCache<PackageStructure>(pkgPath);

      if (cache) {
        if (cache.exists) {
          return cache.content;
        }
        // Fall-through
      } else if (pkgPath.exists()) {
        return this.cache.cacheFileContents(pkgPath, () => loadJson(pkgPath));
      } else {
        this.cache.markMissingFile(pkgPath);
      }

      currentDir = currentDir.parent();
    }

    throw new RuntimeError('config', 'CFG_PACKAGE_SCOPE_UNKNOWN');
  }

  /**
   * Find all configuration and environment specific files in a directory
   * by looping through all the defined extension options.
   * Will only search until the first file is found, and will not return multiple extensions.
   */
  async findFilesInDir(dir: Path): Promise<Path[]> {
    const isRoot = this.isRootDir(dir);
    const baseDir = isRoot ? dir.append(CONFIG_FOLDER) : dir;

    return this.cache.cacheFilesInDir(baseDir, async () => {
      const paths: Path[] = [];

      // eslint-disable-next-line no-restricted-syntax
      for (const ext of this.options.extensions) {
        const files = [baseDir.append(this.getFileName(ext, !isRoot, false))];

        if (this.options.includeEnv) {
          files.push(baseDir.append(this.getFileName(ext, !isRoot, true)));
        }

        await Promise.all(
          files.map(configPath => {
            if (configPath.exists()) {
              paths.push(configPath);
            }

            return configPath;
          }),
        );

        // Once we find any file, we abort looking for others
        if (paths.length > 0) {
          break;
        }
      }

      return paths;
    });
  }

  /**
   * Create and return a config file name, with optional branch and environment variants.
   */
  getFileName(ext: string, isBranch: boolean, isEnv: boolean): string {
    const { name } = this.options;

    return createFileName(name, ext, {
      envSuffix: isEnv ? getEnv(name) : '',
      leadingDot: isBranch,
    });
  }

  /**
   * Load file and package contents from a list of file paths.
   * Extract and apply extended and override configs based on the base path.
   */
  async resolveFiles(basePath: Path, foundFiles: Path[]): Promise<ConfigFile<T>[]> {
    const configs = await Promise.all(foundFiles.map(filePath => this.loadConfig(filePath)));

    // Overrides take the highest precedence and must appear after everything,
    // including branch level configs. However, they must extract first so that
    // extends functionality can be inherited.
    if (this.options.overridesSetting) {
      const overriddenConfigs = await this.extractOverriddenConfigs(basePath, configs);

      if (overriddenConfigs.length > 0) {
        configs.push(...overriddenConfigs);
      }
    }

    // Configs that have been extended from root configs must
    // appear before everything else, in the order they were defined
    if (this.options.extendsSetting) {
      const extendedConfigs = await this.extractExtendedConfigs(configs);

      if (extendedConfigs.length > 0) {
        configs.unshift(...extendedConfigs);
      }
    }

    return configs;
  }

  /**
   * Extract a list of config files to extend, in order, from the list of previously loaded
   * config files, which is typically from the root. The list to extract can be located within
   * a property that matches the `extendsSetting` option.
   */
  protected extractExtendedConfigs(configs: ConfigFile<T>[]): Promise<ConfigFile<T>[]> {
    const { name, extendsSetting } = this.options;
    const extendsPaths: Path[] = [];

    configs.forEach(({ config, path, source }) => {
      const key = extendsSetting as keyof T;
      const extendsFrom = config[key] as ExtendsSetting | undefined;

      if (source === 'root' || source === 'override') {
        delete config[key];
      } else if (extendsFrom) {
        throw new RuntimeError('config', 'CFG_EXTENDS_ROOT_ONLY', [key]);
      } else {
        return;
      }

      toArray(extendsFrom).forEach(extendsPath => {
        if (isModuleName(extendsPath)) {
          extendsPaths.push(
            new Path(extendsPath, createFileName(name, 'js', { envSuffix: 'preset' })),
          );
        } else if (isFilePath(extendsPath)) {
          extendsPaths.push(path.parent().append(extendsPath));
        } else {
          throw new RuntimeError('config', 'CFG_EXTENDS_UNKNOWN_PATH', [extendsPath]);
        }
      });
    });

    return Promise.all(extendsPaths.map(path => this.loadConfig(path)));
  }

  /**
   * Extract all root config overrides that match the current path used to load with.
   * Overrides are located within a property that matches the `overridesSetting` option.
   */
  protected extractOverriddenConfigs(basePath: Path, configs: ConfigFile<T>[]): ConfigFile<T>[] {
    const { overridesSetting } = this.options;
    const overriddenConfigs: ConfigFile<T>[] = [];

    configs.forEach(({ config, path, source }) => {
      const key = overridesSetting as keyof T;
      const overrides = config[key] as OverridesSetting<T>[] | undefined;

      if (source === 'root') {
        delete config[key];
      } else if (overrides) {
        throw new RuntimeError('config', 'CFG_OVERRIDES_ROOT_ONLY', [key]);
      } else {
        return;
      }

      toArray(overrides).forEach(({ exclude, include, settings }) => {
        const excludePatterns = toArray(exclude);
        const includePatterns = toArray(include);
        const options = { dot: true, matchBase: true };

        if (
          includePatterns.some(pattern => minimatch(basePath.path(), pattern, options)) &&
          !excludePatterns.some(pattern => minimatch(basePath.path(), pattern, options))
        ) {
          overriddenConfigs.push({
            config: settings,
            path,
            source: 'override',
          });
        }
      });
    });

    return overriddenConfigs;
  }

  /**
   * Load config contents from the provided file path using one of the defined loaders.
   */
  protected async loadConfig(path: Path): Promise<ConfigFile<T>> {
    const pkg = await this.determinePackageScope(path);
    const config = await this.cache.cacheFileContents(path, async () => {
      const { loaders } = this.options;
      const ext = path.ext(true);

      switch (ext) {
        case 'cjs':
          return loaders.cjs(path, pkg);
        case 'js':
          return loaders.js(path, pkg);
        case 'json':
          return loaders.json(path, pkg);
        case 'mjs':
          // Not easily testable yet
          // istanbul ignore next
          return loaders.mjs(path, pkg);
        case 'yaml':
        case 'yml':
          return loaders.yaml(path, pkg);
        default:
          throw new RuntimeError('config', 'CFG_LOADER_UNSUPPORTED', [ext]);
      }
    });

    return {
      config,
      path,
      source: path.path().includes(CONFIG_FOLDER) ? 'root' : 'branch',
    };
  }
}
