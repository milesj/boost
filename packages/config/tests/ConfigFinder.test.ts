import { Path } from '@boost/common';
import { mockFilePath, normalizeSeparators } from '@boost/common/test';
import { copyFixtureToTempFolder, getFixturePath } from '@boost/test-utils';
import { Cache } from '../src/Cache';
import { ConfigFinder } from '../src/ConfigFinder';
import { mockSystemPath } from './helpers';

describe('ConfigFinder', () => {
	let cache: Cache;
	let finder: ConfigFinder<{ debug: boolean }>;

	beforeEach(() => {
		cache = new Cache();
		finder = new ConfigFinder(
			{
				extendsSetting: 'extends',
				name: 'boost',
				overridesSetting: 'overrides',
			},
			cache,
		);
	});

	it('errors if name is not in camel case', () => {
		expect(() => {
			finder = new ConfigFinder({ name: 'Boosty Boost' }, cache);
		}).toThrow(
			'Invalid ConfigFinder field "name". String must be in camel case. (pattern "^[a-z][0-9A-Za-z]+$")',
		);
	});

	describe('determinePackageScope()', () => {
		it('returns the parent `package.json`', async () => {
			const tempRoot = getFixturePath('config-package-file-tree-monorepo');

			const pkg1 = await finder.determinePackageScope(
				mockFilePath(`${tempRoot}/packages/core/src/index.ts`),
			);

			expect(pkg1).toEqual({ name: 'core' });

			const pkg2 = await finder.determinePackageScope(
				mockFilePath(`${tempRoot}/packages/log/src/index.js`),
			);

			expect(pkg2).toEqual({ name: 'log' });
		});

		it('returns the first parent `package.json` if there are multiple', async () => {
			const tempRoot = getFixturePath('config-package-file-tree-monorepo');

			const pkg = await finder.determinePackageScope(
				mockFilePath(`${tempRoot}/packages/plugin/nested/example/src`),
			);

			expect(pkg).toEqual({ name: 'plugin-example' });
		});

		it('returns root `package.json` if outside a monorepo', async () => {
			const tempRoot = getFixturePath('config-package-file-tree-monorepo');

			const pkg = await finder.determinePackageScope(mockFilePath(`${tempRoot}/index.ts`));

			expect(pkg).toEqual({ name: 'boost-config-package-file-tree-monorepo', version: '0.0.0' });
		});

		it('uses the cache for the same `package.json` parent', async () => {
			const tempRoot = getFixturePath('config-package-file-tree-monorepo');

			const pkg1 = await finder.determinePackageScope(
				mockFilePath(`${tempRoot}/packages/core/src/index.ts`),
			);
			const pkg2 = await finder.determinePackageScope(
				mockFilePath(`${tempRoot}/packages/core/src/deep/nested/core.ts`),
			);

			expect(pkg2).toEqual(pkg1);
		});

		it('caches each depth, even if a file is missing', async () => {
			const tempRoot = getFixturePath('config-package-file-tree-monorepo');

			await finder.determinePackageScope(
				mockFilePath(`${tempRoot}/packages/core/src/deep/nested/core.ts`),
			);

			expect(cache.fileContentCache).toEqual({
				[normalizeSeparators(`${tempRoot}/packages/core/package.json`)]: {
					content: { name: 'core' },
					exists: true,
					mtime: expect.any(Number),
				},
				[normalizeSeparators(`${tempRoot}/packages/core/src/package.json`)]: {
					content: null,
					exists: false,
					mtime: 0,
				},
				[normalizeSeparators(`${tempRoot}/packages/core/src/deep/package.json`)]: {
					content: null,
					exists: false,
					mtime: 0,
				},
				[normalizeSeparators(`${tempRoot}/packages/core/src/deep/nested/package.json`)]: {
					content: null,
					exists: false,
					mtime: 0,
				},
			});
		});

		it('errors if no parent `package.json`', async () => {
			const tempRoot = copyFixtureToTempFolder('config-root-without-package-json');

			await expect(
				finder.determinePackageScope(new Path(`${tempRoot}/nested/index.js`)),
			).rejects.toThrow('Unable to determine package scope. No parent `package.json` found.');
		});
	});

	describe('loadFromBranchToRoot()', () => {
		const fixtures = [
			{ ext: 'js', root: getFixturePath('config-file-tree-js') },
			{ ext: 'json', root: getFixturePath('config-file-tree-json') },
			{ ext: 'json5', root: getFixturePath('config-file-tree-json5') },
			{ ext: 'cjs', root: getFixturePath('config-file-tree-cjs') },
			// { ext: 'ts', root: getFixturePath('config-file-tree-ts')  },
			{ ext: 'yaml', root: getFixturePath('config-file-tree-yaml') },
			{ ext: 'yml', root: getFixturePath('config-file-tree-yml') },
		];

		fixtures.forEach(({ ext, root: tempRoot }) => {
			it(`returns all \`.${ext}\` config files from a branch up to root`, async () => {
				const files = await finder.loadFromBranchToRoot(`${tempRoot}/src/app/profiles/settings`);

				expect(files).toEqual([
					{
						config: { debug: true },
						path: mockSystemPath(`${tempRoot}/.config/boost.json`),
						source: 'root',
					},
					{
						config: { debug: false },
						path: mockSystemPath(`${tempRoot}/src/app/.boost.${ext}`),
						source: 'branch',
					},
					{
						config: { verbose: true },
						path: mockSystemPath(`${tempRoot}/src/app/profiles/settings/.boost.${ext}`),
						source: 'branch',
					},
				]);
			});
		});

		it('returns all config files for all types from a branch up to root', async () => {
			const tempRoot = getFixturePath('config-file-tree-all-types');

			const files = await finder.loadFromBranchToRoot(
				normalizeSeparators(`${tempRoot}/src/app/profiles/settings`),
			);

			expect(files).toEqual([
				{
					config: { debug: true },
					path: mockSystemPath(`${tempRoot}/.config/boost.json`),
					source: 'root',
				},
				{
					config: { type: 'json' },
					path: mockSystemPath(`${tempRoot}/src/.boost.json`),
					source: 'branch',
				},
				{
					config: { type: 'cjs' },
					path: mockSystemPath(`${tempRoot}/src/app/.boost.cjs`),
					source: 'branch',
				},
				{
					config: { type: 'js' },
					path: mockSystemPath(`${tempRoot}/src/app/profiles/.boost.js`),
					source: 'branch',
				},
				{
					config: { type: 'yaml' },
					path: mockSystemPath(`${tempRoot}/src/app/profiles/settings/.boost.yaml`),
					source: 'branch',
				},
			]);
		});

		it('doesnt load config files above root', async () => {
			const tempRoot = getFixturePath('config-scenario-configs-above-root');

			const files = await finder.loadFromBranchToRoot(
				normalizeSeparators(`${tempRoot}/nested/deep`),
			);

			expect(files).toEqual([
				{
					config: { root: true },
					path: mockSystemPath(`${tempRoot}/nested/.config/boost.json`),
					source: 'root',
				},
				{
					config: { aboveRoot: true },
					path: mockSystemPath(`${tempRoot}/nested/deep/.boost.json`),
					source: 'branch',
				},
			]);
		});

		it('doesnt load branch config files that do not start with a period', async () => {
			const tempRoot = getFixturePath('config-scenario-branch-invalid-file-name');

			const files = await finder.loadFromBranchToRoot(normalizeSeparators(`${tempRoot}/src/app`));

			expect(files).toEqual([
				{
					config: { debug: true },
					path: mockSystemPath(`${tempRoot}/.config/boost.json`),
					source: 'root',
				},
			]);
		});

		it('doesnt load branch config files that have an unknown extension', async () => {
			const tempRoot = getFixturePath('config-scenario-branch-invalid-file-name');

			const files = await finder.loadFromBranchToRoot(normalizeSeparators(`${tempRoot}/src/app`));

			expect(files).toEqual([
				{
					config: { debug: true },
					path: mockSystemPath(`${tempRoot}/.config/boost.json`),
					source: 'root',
				},
			]);
		});

		it('doesnt load multiple branch config file types, only the first  found', async () => {
			const tempRoot = getFixturePath('config-scenario-branch-multiple-types');

			const files = await finder.loadFromBranchToRoot(normalizeSeparators(`${tempRoot}/src/app`));

			expect(files).toEqual([
				{
					config: { debug: true },
					path: mockSystemPath(`${tempRoot}/.config/boost.json`),
					source: 'root',
				},
				{
					config: { type: 'json' },
					path: mockSystemPath(`${tempRoot}/src/app/.boost.json`),
					source: 'branch',
				},
			]);
		});

		describe('environment context', () => {
			it('loads branch config files (using BOOST_ENV)', async () => {
				process.env.BOOST_ENV = 'test';

				const tempRoot = getFixturePath('config-scenario-branch-with-envs');

				const files = await finder.loadFromBranchToRoot(normalizeSeparators(`${tempRoot}/src/app`));

				expect(files).toEqual([
					{
						config: { debug: true },
						path: mockSystemPath(`${tempRoot}/.config/boost.json`),
						source: 'root',
					},
					{
						config: { env: 'all' },
						path: mockSystemPath(`${tempRoot}/src/app/.boost.json`),
						source: 'branch',
					},
					{
						config: { env: 'test' },
						path: mockSystemPath(`${tempRoot}/src/app/.boost.test.json`),
						source: 'branch',
					},
				]);

				delete process.env.BOOST_ENV;
			});

			it('loads branch config files (using NODE_ENV)', async () => {
				process.env.NODE_ENV = 'staging';

				const tempRoot = getFixturePath('config-scenario-branch-with-envs');

				const files = await finder.loadFromBranchToRoot(normalizeSeparators(`${tempRoot}/src/app`));

				expect(files).toEqual([
					{
						config: { debug: true },
						path: mockSystemPath(`${tempRoot}/.config/boost.json`),
						source: 'root',
					},
					{
						config: { env: 'all' },
						path: mockSystemPath(`${tempRoot}/src/app/.boost.json`),
						source: 'branch',
					},
					{
						config: { env: 'staging' },
						path: mockSystemPath(`${tempRoot}/src/app/.boost.staging.json`),
						source: 'branch',
					},
				]);

				process.env.NODE_ENV = 'test';
			});

			it('doesnt load branch config files if `includeEnv` is false (using BOOST_ENV)', async () => {
				process.env.BOOST_ENV = 'test';

				const tempRoot = getFixturePath('config-scenario-branch-with-envs');

				finder.configure({ includeEnv: false });

				const files = await finder.loadFromBranchToRoot(normalizeSeparators(`${tempRoot}/src/app`));

				expect(files).toEqual([
					{
						config: { debug: true },
						path: mockSystemPath(`${tempRoot}/.config/boost.json`),
						source: 'root',
					},
					{
						config: { env: 'all' },
						path: mockSystemPath(`${tempRoot}/src/app/.boost.json`),
						source: 'branch',
					},
				]);

				delete process.env.BOOST_ENV;
			});
		});

		describe('overrides', () => {
			it('adds overrides that match the `include` glob', async () => {
				const tempRoot = getFixturePath('config-overrides-from-branch');

				const files = await finder.loadFromBranchToRoot(
					normalizeSeparators(`${tempRoot}/src/foo/does-match.ts`),
				);

				expect(files).toEqual([
					{
						config: { level: 1 },
						path: mockSystemPath(`${tempRoot}/.config/boost.json`),
						source: 'root',
					},
					{
						config: { level: 2 },
						path: mockSystemPath(`${tempRoot}/.config/boost.json`),
						source: 'overridden',
					},
				]);
			});

			it('adds overrides that match the `include` glob and dont match the `exclude` glob', async () => {
				const tempRoot = getFixturePath('config-overrides-from-branch-with-excludes');

				const files = await finder.loadFromBranchToRoot(
					normalizeSeparators(`${tempRoot}/src/bar.ts`),
				);

				expect(files).toEqual([
					{
						config: { level: 1 },
						path: mockSystemPath(`${tempRoot}/.config/boost.json`),
						source: 'root',
					},
					{
						config: { level: 2 },
						path: mockSystemPath(`${tempRoot}/.config/boost.json`),
						source: 'overridden',
					},
				]);
			});

			it('adds overrides using a custom settings name', async () => {
				const tempRoot = getFixturePath('config-overrides-custom-settings-name');

				finder.configure({ overridesSetting: 'rules' });

				const files = await finder.loadFromBranchToRoot(
					normalizeSeparators(`${tempRoot}/src/foo/does-match.ts`),
				);

				expect(files).toEqual([
					{
						config: { level: 1 },
						path: mockSystemPath(`${tempRoot}/.config/boost.json`),
						source: 'root',
					},
					{
						config: { level: 2 },
						path: mockSystemPath(`${tempRoot}/.config/boost.json`),
						source: 'overridden',
					},
				]);
			});

			it('doesnt add overrides that dont match the `include` glob', async () => {
				const tempRoot = getFixturePath('config-overrides-from-branch');

				const files = await finder.loadFromBranchToRoot(
					normalizeSeparators(`${tempRoot}/src/foo/doesnt-match.js`),
				);

				expect(files).toEqual([
					{
						config: { level: 1 },
						path: mockSystemPath(`${tempRoot}/.config/boost.json`),
						source: 'root',
					},
				]);
			});

			it('doesnt add overrides that match the `include` glob AND the `exclude` glob', async () => {
				const tempRoot = getFixturePath('config-overrides-from-branch-with-excludes');

				const files = await finder.loadFromBranchToRoot(
					normalizeSeparators(`${tempRoot}/src/baz.ts`),
				);

				expect(files).toEqual([
					{
						config: { level: 1 },
						path: mockSystemPath(`${tempRoot}/.config/boost.json`),
						source: 'root',
					},
				]);
			});

			it('errors if overrides are found in a branch config', async () => {
				const tempRoot = getFixturePath('config-invalid-branch-nested-overrides');

				await expect(
					finder.loadFromBranchToRoot(normalizeSeparators(`${tempRoot}/src`)),
				).rejects.toThrow('Overrides setting `overrides` must only be defined in a root config.');
			});
		});

		describe('extends', () => {
			it('extends configs using relative and absolute file paths', async () => {
				const tempRoot = getFixturePath('config-extends-fs-paths');

				const files = await finder.loadFromBranchToRoot(tempRoot);

				expect(files).toEqual([
					{
						config: { relative: true },
						path: mockSystemPath(`${tempRoot}/some/relative/path/config.js`),
						source: 'extended',
					},
					{
						config: { absolute: true },
						path: mockSystemPath(`${tempRoot}/some/absolute/path/config.yml`, false),
						source: 'extended',
					},
					{
						config: { root: true },
						path: mockSystemPath(`${tempRoot}/.config/boost.js`),
						source: 'root',
					},
				]);
			});

			it('extends config presets from node modules', async () => {
				const tempRoot = copyFixtureToTempFolder('config-extends-module-presets');

				finder.configure({
					resolver: (id) => normalizeSeparators(`${tempRoot}/node_modules/${id}`),
				});

				const files = await finder.loadFromBranchToRoot(tempRoot);

				expect(files).toEqual([
					{
						config: { name: 'foo' },
						path: mockSystemPath(`${tempRoot}/node_modules/foo/boost.preset.js`),
						source: 'extended',
					},
					{
						config: { name: '@scope/bar' },
						path: mockSystemPath(`${tempRoot}/node_modules/@scope/bar/boost.preset.js`),
						source: 'extended',
					},
					{
						config: { root: true },
						path: mockSystemPath(`${tempRoot}/.config/boost.json`),
						source: 'root',
					},
				]);
			});

			it('extends configs that were defined in an override', async () => {
				const tempRoot = getFixturePath('config-extends-from-override');

				const files = await finder.loadFromBranchToRoot(
					normalizeSeparators(`${tempRoot}/some/relative/path`),
				);

				expect(files).toEqual([
					{
						config: { extended: true },
						path: mockSystemPath(`${tempRoot}/some/relative/path/config.js`),
						source: 'extended',
					},
					{
						config: { root: true },
						path: mockSystemPath(`${tempRoot}/.config/boost.json`),
						source: 'root',
					},
					{
						config: { overridden: true },
						path: mockSystemPath(`${tempRoot}/.config/boost.json`),
						source: 'overridden',
					},
				]);
			});

			it('extends configs using a custom settings name', async () => {
				const tempRoot = copyFixtureToTempFolder('config-extends-custom-setting-name');

				finder.configure({
					resolver: (id) => normalizeSeparators(`${tempRoot}/node_modules/${id}`),
				});

				finder.configure({ extendsSetting: 'presets' });

				const files = await finder.loadFromBranchToRoot(tempRoot);

				expect(files).toEqual([
					{
						config: { type: 'module' },
						path: mockSystemPath(`${tempRoot}/node_modules/foo/boost.preset.js`),
						source: 'extended',
					},
					{
						config: { type: 'fs' },
						path: mockSystemPath(`${tempRoot}/some/relative/path/config.js`),
						source: 'extended',
					},
					{
						config: { root: true },
						path: mockSystemPath(`${tempRoot}/.config/boost.json`),
						source: 'root',
					},
				]);
			});

			it('errors if extends are found in a branch config', async () => {
				const tempRoot = getFixturePath('config-invalid-branch-nested-extends');

				await expect(
					finder.loadFromBranchToRoot(normalizeSeparators(`${tempRoot}/src`)),
				).rejects.toThrow('Extends setting `extends` must only be defined in a root config.');
			});

			it('errors for an invalid extends path', async () => {
				const tempRoot = getFixturePath('config-invalid-extends-path');

				await expect(finder.loadFromBranchToRoot(tempRoot)).rejects.toThrow(
					'Cannot extend configuration. Unknown module or file path "123!#?".',
				);
			});

			it('errors for a missing extends path', async () => {
				const tempRoot = getFixturePath('config-missing-extends-path');

				await expect(finder.loadFromBranchToRoot(tempRoot)).rejects.toThrow(
					'no such file or directory',
				);
			});
		});
	});

	describe('loadFromRoot()', () => {
		const fixtures = [
			{ ext: 'js', root: getFixturePath('config-root-config-js') },
			{ ext: 'json', root: getFixturePath('config-root-config-json') },
			{ ext: 'json5', root: getFixturePath('config-root-config-json5') },
			{ ext: 'cjs', root: getFixturePath('config-root-config-cjs') },
			// { ext: 'ts', root: getFixturePath('config-root-config-ts') },
			{ ext: 'yaml', root: getFixturePath('config-root-config-yaml') },
			{ ext: 'yml', root: getFixturePath('config-root-config-yml') },
		];

		fixtures.forEach(({ ext, root: tempRoot }) => {
			it(`returns \`.${ext}\` config file from root`, async () => {
				const files = await finder.loadFromRoot(tempRoot);

				expect(files).toEqual([
					{
						config: { debug: true },
						path: mockSystemPath(`${tempRoot}/.config/boost.${ext}`),
						source: 'root',
					},
				]);
			});
		});

		it('errors if not root folder', async () => {
			const tempRoot = getFixturePath('config-scenario-not-root');

			await expect(finder.loadFromRoot(tempRoot)).rejects.toThrow(
				'Invalid configuration root. Requires a `.config` folder and `package.json`.',
			);
		});

		it('errors if root folder is missing a `package.json`', async () => {
			const tempRoot = getFixturePath('config-root-without-package-json');

			await expect(finder.loadFromRoot(tempRoot)).rejects.toThrow(
				'Config folder `.config` found without a relative `package.json`. Both must be located in the project root.',
			);
		});

		it('errors for invalid config file/loader type', async () => {
			const tempRoot = getFixturePath('config-root-config-toml');

			finder.configure(
				// @ts-expect-error Invalid format
				{ extensions: ['toml'] },
			);

			await expect(finder.loadFromRoot(tempRoot)).rejects.toThrow(
				'Unsupported loader format "toml".',
			);
		});
	});
});
