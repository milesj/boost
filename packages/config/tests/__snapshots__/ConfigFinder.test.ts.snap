// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigFinder errors if name is not in camel case 1`] = `
"The following validations have failed for \`ConfigFinder\`:
  - Invalid field \\"name\\". String must be in camel case. (pattern \\"^[a-z][a-zA-Z0-9]+$\\")"
`;
